bpf_host: emit '-> network' traces for egress packets

navarrothiago opened the pull request:
bpf_host: emit '-> network' traces for egress packets

See the issue below for the full context. The caveat is, this program is only attached to the native device egress if the host firewall is enabled, so there still won't be any TRACE_TO_NETWORK visibility if that's not the case.

I've also added handle_to_netdev_ipv4() to clean up to_netdev() a bit, to bring it on par with the ipv6 handling.

bpf_host: emit '-> network' traces for packets leaving the node when Host Firewall are enabled

Based on: #13485
Fixes: #12561

Co-authored-by: Thiago Navarro <navarro@accuknox.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Thiago Navarro <navarro@accuknox.com>

pchaigno commented on bpf/lib/common.h:
```diff
@@ -563,31 +563,37 @@ static __always_inline __u32 or_encrypt_key(__u8 key)
+#define	CB_ADDR_V6_4		CB_CT_STATE
+#define	CB_ENCRYPT_DST		CB_CT_STATE
+#define	CB_CUSTOM_CALLS		CB_CT_STATE
+#define	CB_CT_MONITOR		CB_CT_STATE
```
What happened in this file?

pchaigno commented on bpf/lib/trace.h:
```diff
@@ -141,6 +141,13 @@ static __always_inline bool emit_trace_notify(__u8 obs_point, __u32 monitor)
+	 * so if there is no CT information available at the observation point,
+	 * then 'monitor' will be set to 0 to avoid emitting trace notifications
+	 * when aggregation is enabled (the default).
+	 */
```
This sort of cleanup should be in a separate commit.

pchaigno reviewed (commented):
This is a bit hard to review as is. Could you separate in multiple logical commits? For example, the changes to take monitor into account in existing traces (what #12561 is about) could be one commit, new traces would be another commit, and general cleanups (e.g., new comments) would be a last commit.

ti-mo committed:
bpf_host: emit '-> network' traces for egress packets

See the issue below for the full context. The caveat is, this program is
only attached to the native device egress if the host firewall is
enabled, so there still won't be any TRACE_TO_NETWORK visibility if
that's not the case.

bpf_host: emit '-> network' traces for packets leaving the node when
Host Firewall are enabled

Based on: #13485
Fixes: #12561

Co-authored-by: Thiago Navarro <navarro@accuknox.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Thiago Navarro <navarro@accuknox.com>

ti-mo committed:
bpf: adds handle_to_netdev_ipv4() to clean up to_netdev()

Besides, cleans up with some comments and applies code style

Co-authored-by: Thiago Navarro <navarro@accuknox.com>
Signed-off-by: Timo Beckers <timo@isovalent.com>
Signed-off-by: Thiago Navarro <navarro@accuknox.com>

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1217,13 +1222,17 @@ static __always_inline int nodeport_nat_ipv4_fwd(struct __ctx_buff *ctx)
 	if (ret == NAT_PUNT_TO_STACK)
 		ret = CTX_ACT_OK;
 
+	send_trace_notify(ctx, TRACE_TO_NETWORK, 0, 0, 0, 0, ret, monitor);
```
I think we should skip the commit that adds `TRACE_TO_NETWORK` traces for now. See the discussion at https://github.com/cilium/cilium/pull/14696. These additional traces will get very noisy quickly so there's a bit more work needed before we can introduce them.

pchaigno reviewed (changes requested):
I think we should exclude the second commit for now. The other two commits look :+1: to me.

Please don't forget to add Timo has a coauthor of the last commit (via `Signed-off-by`, `Co-authored-by`, or both).
